### DISCUSSION TITLE
How to fail `SpringBootApplication` fail on unknown Kafka topicname, group-id and such?

### DISCUSSION THREAD

**Discussion by JooHyukKim** [spring-projects/spring-kafka#3002] [Q&A] [answered]
I am trying to make the server fail on start-up.

I was wondering if there is any config that does such thing?

**Comment by JooHyukKim** [comment, marked as answer]
Found workaround.

- Made my @Compontent with `@KafkaListener` annotoation to implement ApplicationRunner.
- in run command, validates existent of needed topic by the compontent itself.
- WARNING : quite verbose, but trying to not let the sever run unintneded.

```
/**
     * At the start of the Application, this method verifies the connection to Kafka and the existence of a specific topic.
     */
    @Throws(Exception::class)
    override fun run(args: ApplicationArguments) {
        // Creating an Admin Client
        val config = Properties()
        // Setting the address of the Kafka broker
        config[AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG] = bootstrapServers
        // Establishing a connection to Kafka
        val adminClient = AdminClient.create(config)
        // Retrieving a list of all topics from Kafka
        val topics = adminClient.listTopics().names().get()
        // Validating the existence of the specific topic
        if (!topics.contains(topicToCheck)) {
            // If the topic is not found, throw an exception and terminate the application
            throw IllegalStateException("Topic $topicToCheck not found. Application will exit.")
        }
```

**Reply by artembilan** [reply]
There is also Kafka broker property to not create topic automatically: `auto.create.topics.enable=false`.
This way you application is going to fail trying to interact with non-existing topics.

**Reply by Casper-dss** [reply]
I have set the auto.create.topics.enable=false in kafka. But my spring boot application keeps logging this warning.

{virksomhed-data-request-retry=UNKNOWN_TOPIC_OR_PARTITION, virksomhed-data-request=UNKNOWN_TOPIC_OR_PARTITION}

Is ot possible to make the service stop, after ex. 5 retries?

**Reply by Casper-dss** [reply]
ChatGPT was my freind. This works. Define an admin bean

```
    @Bean
    public AdminClient adminClient(ConsumerFactory<String, Object> factory) {
        return AdminClient.create(factory.getConfigurationProperties());
    }
```
 And then use the bean in a PostConstruct in ex. Application class
```
    @Autowired
    private AdminClient adminClient;

    @PostConstruct
    public void verifyTopicExists() {
        try {
            ListTopicsResult listTopics = adminClient.listTopics();
            Set<String> topics = listTopics.names().get(10, TimeUnit.SECONDS); // Adjust the timeout as necessary
            if (!topics.contains(Topics.VIRKSOMHED_DATA_REQUEST)) { // Replace Topics.VIRKSOMHED_DATA_REQUEST with the name of your topic
                throw new IllegalStateException("Required Kafka topic does not exist");
            }
        } catch (Exception e) {
            throw new RuntimeException("Failed to verify Kafka topic existence", e);
        }
    }
```
This works for me. The example is straight from ChatGPT, so the exception handling can be done more smoothly. It is kind of the same as JooHyukKim suggested.

**Reply by JooHyukKim** [reply]
> There is also Kafka broker property to not create topic automatically: `auto.create.topics.enable=false`. This way you application is going to fail trying to interact with non-existing topics.

This couldn't do because my requirement was that Application should fail to start up if Topic does not exist.
